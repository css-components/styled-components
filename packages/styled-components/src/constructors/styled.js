// @flow
import constructWithOptions from './constructWithOptions';
import StyledComponent from '../models/StyledComponent';
import domElements from '../utils/domElements';

import type { Target } from '../types';

const styled = (tag: Target) => constructWithOptions(StyledComponent, tag);

// Shorthands for all valid HTML Elements
for (let i = 0; i < domElements.length; i += 1) {
  styled[domElements[i]] = styled(domElements[i]);
}

export default styled;
