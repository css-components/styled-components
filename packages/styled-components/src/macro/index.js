// @flow
import traverse from '@babel/traverse';
import { createMacro } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';

function styledComponentsMacro({ references, source, state, babel: { types }, config = {} }) {
  // Replace macro imports with `styled-components`.
  traverse(state.file.ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === source) {
        path.node.source.value = 'styled-components';
      }
    },
    VariableDeclaration(path) {
      path.node.declarations.forEach(declaration => {
        if (declaration.init.type === 'CallExpression') {
          const { callee, arguments: args } = declaration.init;
          if (
            callee.type === 'Identifier' &&
            callee.name === 'require' &&
            args.length === 1 &&
            args[0].value === source
          ) {
            args[0].value = 'styled-components';
          }
        }
      });
    },
  });

  // Find name of default export if it was used.
  let customImportName;
  if ('default' in references && references['default'].length > 0) {
    customImportName = references['default'][0].node.name;
  }

  // Run Babel plugin.
  const stateWithOpts = { ...state, opts: config, customImportName };
  traverse(state.file.ast, babelPlugin({ types }).visitor, undefined, stateWithOpts);

  // Don't remove the newly-modified import statement.
  return { keepImports: true };
}

export default createMacro(styledComponentsMacro, {
  configName: 'styledComponents',
});
