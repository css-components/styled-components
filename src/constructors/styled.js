// @flow
import css from './css'
import type { Interpolation, Target } from '../types'
import domElements from '../utils/domElements'
import enforceArray from '../utils/enforce-array-arguments'

export default (styledComponent: Function) => {
  const styled = (tag: Target) =>
    (strings: Array<string>, ...interpolations: Array<Interpolation>) => {
      const tagName = typeof tag === 'string' ? `.${tag}` : '(Component)'
      enforceArray(`styled${tagName}`)(strings)
      return styledComponent(tag, css(strings, ...interpolations))
    }

  // Shorthands for all valid HTML Elements
  domElements.forEach(domElement => {
    styled[domElement] = styled(domElement)
  })

  return styled
}
