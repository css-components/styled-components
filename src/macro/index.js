// @flow
import { createMacro } from 'babel-macros'
import minify from 'babel-plugin-styled-components/lib/visitors/minify'
import displayNameAndId from 'babel-plugin-styled-components/lib/visitors/displayNameAndId'
import templateLiteral from 'babel-plugin-styled-components/lib/visitors/templateLiterals'

const configName = 'styledComponents'

function styledComponentsMacro({ references, state, babel, config = {} }) {
  const t = babel.types
  // add 'import styled from styled-components'
  const literal = config.preprocess ? 'styled-components/no-parser' : 'styled-components'
  state.file.path.parent.program.body.unshift(
    t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier('styled'))],
      t.stringLiteral(literal),
    ),
  )
  // iterate over each call and modify the TaggedTemplateExpression
  references.default.forEach(referencePath => {
    const templatePath = referencePath.parentPath.parentPath
    const basePath = referencePath
    basePath.node.name = 'styled'
    const newState = {
      ...state,
      opts: config,
    }
    if (templatePath.type === 'TaggedTemplateExpression') {
      minify(templatePath, newState)
      displayNameAndId(templatePath, newState)
      templateLiteral(templatePath, newState)
    }
  })
}

module.exports = createMacro(styledComponentsMacro, { configName })
