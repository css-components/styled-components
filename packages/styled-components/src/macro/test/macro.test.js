import cosmiconfigMock from 'cosmiconfig';
import babel from '@babel/core';
import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

jest.mock('cosmiconfig', () => jest.fn(jest.requireActual('cosmiconfig')));

const styledExampleCode = `
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const customStyledExampleCode = `
import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const cssExampleCode = `
import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`
`;

const keyframesExampleCode = `
import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`
`;

const createGlobalStyleExampleCode = `
import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`
`;

const ThemeProviderExampleCode = `
import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider,
  { theme: { color: 'red' }},
  'hello'
)
`;

const extendsExampleCode = `
import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`
`;

const multipleImportsExampleCode = `
import styled, { css } from '../../macro'
export { styled, css }
`;

const requireExampleCode = `
const styled = require('../../macro')

styled.div\`
  background: red;
\`
`;

const withTypeImportExampleCode = `
import { DefaultTheme } from '../../macro'
export { DefaultTheme }
`;

const withTypeAndStandardImportExampleCode = `
import styled, { DefaultTheme } from '../../macro'
export { styled, DefaultTheme }
`;

const cssPropExampleCode = `
import styled from '../../macro'
import React from 'react'
function Foo() {
  return <div css="color: red;" />
}
`;

const cssPropOverridingComponentExampleCode = `
import React from 'react'
import styled from '../../macro'
const Thing = styled.div\`
  color: blue;
\`;
function Foo() {
  return <Thing css="color: red;" />
}
`;

const cssPropStaticCssHelperExampleCode = `
import React from 'react'
import { css } from '../../macro'
function Foo() {
  return <div css={css\`color: red;\`} />
}
`;

const cssPropDynamicCssHelperExampleCode = `
import React from 'react'
import { css } from '../../macro'
const padding = '10px'
function Foo(props) {
  return <div css={css\`
    color: \${props.color};
    font-size: \${p => p.theme.fontSize};
    padding: \${padding};
  \`} />
}
`;

const cssPropMultipleCssHelperExampleCode = `
import React from 'react'
import { css } from '../../macro'
function Foo() {
  return <div css={css\`color: red;\`} />
}
function Bar() {
  return <div css={css\`color: green;\`} />
}
`;

pluginTester({
  title: 'macro',
  plugin,
  snapshot: true,
  babelOptions: {
    babelrc: false,
    filename: __filename,
    presets: ['@babel/react'],
  },
  babel,
  tests: {
    'should work with styled': {
      code: styledExampleCode,
    },
    'should work with custom import name': {
      code: customStyledExampleCode,
    },
    'should work with { css }': {
      code: cssExampleCode,
    },
    'should work with { keyframes }': {
      code: keyframesExampleCode,
    },
    'should work with { createGlobalStyle }': {
      code: createGlobalStyleExampleCode,
    },
    'should work with { ThemeProvider }': {
      code: ThemeProviderExampleCode,
    },
    'should work when extending a component': {
      code: extendsExampleCode,
    },
    'should work with require() to import styled-components': {
      code: requireExampleCode,
    },
    'should work with multiple imports': {
      code: multipleImportsExampleCode,
    },
    'should work with types': {
      code: withTypeImportExampleCode,
    },
    'should work with types alongside import': {
      code: withTypeAndStandardImportExampleCode,
    },
    'should not add componentId with a config disabling ssr': {
      code: styledExampleCode,
      setup: () => {
        cosmiconfigMock.mockImplementationOnce(() => ({
          searchSync: () => ({
            config: {
              styledComponents: {
                ssr: false,
              },
            },
          }),
        }));
      },
    },
    'should work with the css prop': { code: cssPropExampleCode },
    'should work with the css prop overriding an existing styled-component': {
      code: cssPropOverridingComponentExampleCode,
    },
    'should work with the css prop with a static css helper': {
      code: cssPropStaticCssHelperExampleCode,
    },
    'should work with the css prop with a dynamic css helper': {
      code: cssPropDynamicCssHelperExampleCode,
    },
    'should work with the css prop with css helpers on multiple components': {
      code: cssPropMultipleCssHelperExampleCode,
    },
  },
});
