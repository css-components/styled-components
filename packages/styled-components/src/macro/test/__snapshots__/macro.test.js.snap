// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro should not add componentId with a config disabling ssr: should not add componentId with a config disabling ssr 1`] = `
"
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from \\"styled-components\\";
styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:\\", \\";\\"], p => p.error ? 'red' : 'green');
"
`;

exports[`macro should work when extending a component: should work when extending a component 1`] = `
"
import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from \\"styled-components\\";
import React from 'react';

const Hello = () => React.createComponent(div, null, 'hello');

styled(Hello).withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:red;\\"]);
"
`;

exports[`macro should work with { ThemeProvider }: should work with { ThemeProvider } 1`] = `
"
import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider,
  { theme: { color: 'red' }},
  'hello'
)

      ↓ ↓ ↓ ↓ ↓ ↓

import { ThemeProvider } from \\"styled-components\\";
React.createComponent(ThemeProvider, {
  theme: {
    color: 'red'
  }
}, 'hello');
"
`;

exports[`macro should work with { createGlobalStyle }: should work with { createGlobalStyle } 1`] = `
"
import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { createGlobalStyle } from \\"styled-components\\";
createGlobalStyle\`
  background: red;
\`;
"
`;

exports[`macro should work with { css }: should work with { css } 1`] = `
"
import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from \\"styled-components\\";
css([\\"color:\\", \\";\\"], props => props.whiteColor ? 'white' : 'black');
"
`;

exports[`macro should work with { keyframes }: should work with { keyframes } 1`] = `
"
import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { keyframes } from \\"styled-components\\";
keyframes([\\"0%{opacity:0;}100%{opacity:1;}\\"]);
"
`;

exports[`macro should work with custom import name: should work with custom import name 1`] = `
"
import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import myStyled from \\"styled-components\\";
import styled from 'another-package';
myStyled.div\`
  background: \${p => p.error ? 'red' : 'green'};
\`;
"
`;

exports[`macro should work with multiple imports: should work with multiple imports 1`] = `
"
import styled, { css } from '../../macro'
export { styled, css }

      ↓ ↓ ↓ ↓ ↓ ↓

import styled, { css } from \\"styled-components\\";
export { styled, css };
"
`;

exports[`macro should work with require() to import styled-components: should work with require() to import styled-components 1`] = `
"
const styled = require('../../macro')

styled.div\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from \\"styled-components\\";
styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:red;\\"]);
"
`;

exports[`macro should work with styled: should work with styled 1`] = `
"
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from \\"styled-components\\";
styled.div.withConfig({
  displayName: \\"macrotest\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"background:\\", \\";\\"], p => p.error ? 'red' : 'green');
"
`;

exports[`macro should work with the css prop overriding an existing styled-component: should work with the css prop overriding an existing styled-component 1`] = `
"
import React from 'react'
import styled from '../../macro'
const Thing = styled.div\`
  color: blue;
\`;
function Foo() {
  return <Thing css=\\"color: red;\\" />
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import styled from \\"styled-components\\";
import React from 'react';
const Thing = styled.div.withConfig({
  displayName: \\"macrotest__Thing\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"color:blue;\\"]);

var _StyledThing = styled(Thing).withConfig({
  displayName: \\"macrotest___StyledThing\\",
  componentId: \\"sc-11gvsec-1\\"
})([\\"color:red;\\"]);

function Foo() {
  return /*#__PURE__*/React.createElement(_StyledThing, null);
}
"
`;

exports[`macro should work with the css prop with a dynamic css helper: should work with the css prop with a dynamic css helper 1`] = `
"
import React from 'react'
import { css } from '../../macro'
const padding = '10px'
function Foo(props) {
  return <div css={css\`
    color: \${props.color};
    font-size: \${p => p.theme.fontSize};
    padding: \${padding};
  \`} />
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import { css } from \\"styled-components\\";
import React from 'react';
const padding = '10px';

var _StyledDiv = _styled(\\"div\\").withConfig({
  displayName: \\"macrotest___StyledDiv\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"color:\\", \\";font-size:\\", \\";padding:\\", \\";\\"], p => p._css, p => p.theme.fontSize, padding);

function Foo(props) {
  return /*#__PURE__*/React.createElement(_StyledDiv, {
    _css: props.color
  });
}
"
`;

exports[`macro should work with the css prop with a static css helper: should work with the css prop with a static css helper 1`] = `
"
import React from 'react'
import { css } from '../../macro'
function Foo() {
  return <div css={css\`color: red;\`} />
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import { css } from \\"styled-components\\";
import React from 'react';

var _StyledDiv = _styled(\\"div\\").withConfig({
  displayName: \\"macrotest___StyledDiv\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"color:red;\\"]);

function Foo() {
  return /*#__PURE__*/React.createElement(_StyledDiv, null);
}
"
`;

exports[`macro should work with the css prop with css helpers on multiple components: should work with the css prop with css helpers on multiple components 1`] = `
"
import React from 'react'
import { css } from '../../macro'
function Foo() {
  return <div css={css\`color: red;\`} />
}
function Bar() {
  return <div css={css\`color: green;\`} />
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled2 from \\"styled-components\\";
import _styled from \\"styled-components\\";
import { css } from \\"styled-components\\";
import React from 'react';

var _StyledDiv = _styled(\\"div\\").withConfig({
  displayName: \\"macrotest___StyledDiv\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"color:red;\\"]);

function Foo() {
  return /*#__PURE__*/React.createElement(_StyledDiv, null);
}

var _StyledDiv2 = _styled(\\"div\\").withConfig({
  displayName: \\"macrotest___StyledDiv2\\",
  componentId: \\"sc-11gvsec-1\\"
})([\\"color:green;\\"]);

function Bar() {
  return /*#__PURE__*/React.createElement(_StyledDiv2, null);
}
"
`;

exports[`macro should work with the css prop: should work with the css prop 1`] = `
"
import styled from '../../macro'
import React from 'react'
function Foo() {
  return <div css=\\"color: red;\\" />
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"styled-components\\";
import \\"styled-components\\";
import React from 'react';

var _StyledDiv = _styled(\\"div\\").withConfig({
  displayName: \\"macrotest___StyledDiv\\",
  componentId: \\"sc-11gvsec-0\\"
})([\\"color:red;\\"]);

function Foo() {
  return /*#__PURE__*/React.createElement(_StyledDiv, null);
}
"
`;

exports[`macro should work with types alongside import: should work with types alongside import 1`] = `
"
import styled, { DefaultTheme } from '../../macro'
export { styled, DefaultTheme }

      ↓ ↓ ↓ ↓ ↓ ↓

import styled, { DefaultTheme } from \\"styled-components\\";
export { styled, DefaultTheme };
"
`;

exports[`macro should work with types: should work with types 1`] = `
"
import { DefaultTheme } from '../../macro'
export { DefaultTheme }

      ↓ ↓ ↓ ↓ ↓ ↓

import { DefaultTheme } from \\"styled-components\\";
export { DefaultTheme };
"
`;
